cmd/goal/tealsign.go

```diff
@@ -0,0 +1,223 @@
+		/*
+		 * First, fetch the key for signing
+		 */
+
+		if keyFilename != "" && signerAcct != "" {
+			reportErrorf(tealsignMutKeyArgs)
+		}
+
+		var kdata []byte
+		var err error
+		if keyFilename != "" {
+			kdata, err = ioutil.ReadFile(keyFilename)
+			if err != nil {
+				reportErrorf(tealsignKeyfileFail, err)
+			}
+		}
+
+		// --account not yet supported, coming in another PR
+		// (need to add kmd support for signing logicsig data)
+		if signerAcct != "" {
+			reportErrorf(tealsignNoWithAcct)
+		}
+
+		// Create signature secrets from the seed
+		var seed crypto.Seed
+		copy(seed[:], kdata)
+		sec := crypto.GenerateSignatureSecrets(seed)
+
+		/*
+		 * Next, fetch the hash of the program for use in the domain
+		 * separated signature payload
+		 */
+
+		var lsigHashArgs int
+		if lsigTxnFilename != "" {
+			lsigHashArgs++
+		}
+		if contractAddr != "" {
+			lsigHashArgs++
+		}
+
+		// Ensure there is one unambiguous source of program hash
+		if lsigHashArgs != 1 {
+			reportErrorf(tealsignMutLsigArgs)
+		}
+
+		var progHash crypto.Digest
+		var stxn transactions.SignedTxn
+		if lsigTxnFilename != "" {
+			// If passed a SignedTxn with a logic sig, compute
+			// the hash of the program within the logic sig
+			stxnBytes, err := ioutil.ReadFile(lsigTxnFilename)
+			if err != nil {
+				reportErrorf(fileReadError, lsigTxnFilename, err)
+			}
+
+			err = protocol.Decode(stxnBytes, &stxn)
+			if err != nil {
+				reportErrorf(txDecodeError, lsigTxnFilename, err)
+			}
+
+			// Ensure signed transaction has a logic sig with a
+			// program
+			if len(stxn.Lsig.Logic) == 0 {
+				reportErrorf(tealsignEmptyLogic)
+			}
+
+			progHash = crypto.HashObj(logic.Program(stxn.Lsig.Logic))
+		} else {
+			// Otherwise, the contract address is the logic hash
+			parsedAddr, err := basics.UnmarshalChecksumAddress(contractAddr)
+			if err != nil {
+				reportErrorf(tealsignParseAddr, err)
+			}
+
+			// Copy parsed address as program hash
+			copy(progHash[:], parsedAddr[:])
+		}
+
+		/*
+		 * Next, fetch the data to sign
+		 */
+
+		var dataArgs int
+		var dataToSign []byte
+
+		if dataFile != "" {
+			dataToSign, err = ioutil.ReadFile(dataFile)
+			if err != nil {
+				reportErrorf(tealsignParseData, err)
+			}
+			dataArgs++
+		}
+		if datab64 != "" {
+			dataToSign, err = base64.StdEncoding.DecodeString(datab64)
+			if err != nil {
+				reportErrorf(tealsignParseb64, err)
+			}
+			dataArgs++
+		}
+		if datab32 != "" {
+			dataToSign, err = base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(datab32)
+			if err != nil {
+				reportErrorf(tealsignParseb32, err)
+			}
+			dataArgs++
+		}
+		if signTxID {
+			if lsigTxnFilename == "" {
+				reportErrorf(tealsignTxIDLsigReq)
+			}
+			txid := stxn.Txn.ID()
+			dataToSign = txid[:]
+			dataArgs++
+		}
+
+		// Ensure there is one unambiguous source of data
+		if dataArgs != 1 {
+			reportErrorf(tealsignDataReq)
+		}
+
+		/*
+		 * Sign the payload and print signature to stdout
```

Comment:
Nit: remove "and print signature to stdout"

Reply:
done